mucommander-core/src/main/java/com/mucommander/ui/action/impl/ShowTerminalAction.java

```diff
@@ -0,0 +1,140 @@
+/*
+ * This file is part of muCommander, http://www.mucommander.com
+ *
+ * muCommander is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * muCommander is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package com.mucommander.ui.action.impl;
+
+import com.jediterm.terminal.ui.JediTermWidget;
+import com.mucommander.commons.file.AbstractFile;
+import com.mucommander.commons.file.protocol.local.LocalFile;
+import com.mucommander.core.desktop.DesktopManager;
+import com.mucommander.text.Translator;
+import com.mucommander.ui.action.AbstractActionDescriptor;
+import com.mucommander.ui.action.ActionCategory;
+import com.mucommander.ui.action.ActionDescriptor;
+import com.mucommander.ui.action.ActionProperties;
+import com.mucommander.ui.main.MainFrame;
+
+import java.awt.BorderLayout;
+import java.awt.event.KeyAdapter;
+import java.awt.event.KeyEvent;
+import java.util.Map;
+
+import javax.swing.KeyStroke;
+import javax.swing.SwingUtilities;
+
+/**
+ * This action shows built-in terminal (it mimics the behavior of Midnight Commander Ctrl-O command
+ * that originates back from Norton Commander).
+ */
+public class ShowTerminalAction extends ActiveTabAction {
+
+    private JediTermWidget terminal;
+
+    public ShowTerminalAction(MainFrame mainFrame, Map<String, Object> properties) {
+        super(mainFrame, properties);
+
+        setEnabled(DesktopManager.canOpenInFileManager());
+    }
+
+    @Override
+    protected void toggleEnabledState() {
+        AbstractFile currentFolder = mainFrame.getActivePanel().getCurrentFolder();
+        setEnabled(currentFolder.getURL().getScheme().equals(LocalFile.SCHEMA));
+    }
+
+    @Override
+    public void performAction() {
+        if (terminal == null) {
+            try {
+                LOGGER.info("Going to show Terminal...");
+                mainFrame.getSplitPane().setVisible(false);
+                terminal = getTerminal();
+                mainFrame.getMainPanel().add(terminal, BorderLayout.CENTER);
+                terminal.revalidate();
+                SwingUtilities.invokeLater(() -> {
+                    terminal.requestFocusInWindow();
+                });
+
+                terminal.getTerminalPanel().addCustomKeyListener(new KeyAdapter() {
+                    public void keyPressed(KeyEvent keyEvent) {
+                        // TODO get the real configured key seq
```

Comment:
+1

Reply:
do you know whether such a trickery is already being done somewhere in the code?

Reply:
not from the top of my head, I'll need to check